x-pack/plugins/encrypted_saved_objects/server/lib/encrypted_saved_objects_client_wrapper.ts

```diff
@@ -0,0 +1,189 @@
+    // NodeJS thread pool. If it turns out to be a problem, we can consider switching to the
+    // sequential processing.
+    const encryptedObjects = await Promise.all(
+      objects.map(async object => {
+        if (!this.options.service.isRegistered(object.type)) {
+          return object;
+        }
+
+        // Saved objects with encrypted attributes should have IDs that are hard to guess especially
+        // since IDs are part of the AAD used during encryption, that's why we control them within this
+        // wrapper and don't allow consumers to specify their own IDs directly.
+        if (object.id) {
+          throw new Error(
+            'Predefined IDs are not allowed for saved objects with encrypted attributes.'
+          );
+        }
+
+        const id = generateID();
+        return {
+          ...object,
+          id,
+          attributes: await this.options.service.encryptAttributes(
+            object.type,
+            id,
+            object.attributes
+          ),
+        };
+      })
+    );
+
+    return this.stripEncryptedAttributesFromBulkResponse(
+      await this.options.baseClient.bulkCreate(encryptedObjects, options)
+    );
+  }
+
+  public async delete(type: string, id: string, options?: BaseOptions) {
+    return await this.options.baseClient.delete(type, id, options);
+  }
+
+  public async find(options: FindOptions = {}) {
+    return this.stripEncryptedAttributesFromBulkResponse(
+      await this.options.baseClient.find(options)
+    );
+  }
+
+  public async bulkGet(objects: BulkGetObjects = [], options?: BaseOptions) {
+    return this.stripEncryptedAttributesFromBulkResponse(
+      await this.options.baseClient.bulkGet(objects, options)
+    );
+  }
+
+  public async get(type: string, id: string, options?: BaseOptions) {
+    return this.stripEncryptedAttributesFromResponse(
+      await this.options.baseClient.get(type, id, options)
+    );
+  }
+
+  public async update<T extends SavedObjectAttributes>(
+    type: string,
+    id: string,
+    attributes: Partial<T>,
+    options?: UpdateOptions
+  ) {
+    if (!this.options.service.isRegistered(type)) {
+      return await this.options.baseClient.update(type, id, attributes, options);
+    }
+
+    return this.stripEncryptedAttributesFromResponse(
+      await this.options.baseClient.update(
+        type,
+        id,
+        await this.options.service.encryptAttributes(type, id, attributes),
+        options
+      )
+    );
+  }
+
+  /**
+   * Strips encrypted attributes from any non-bulk Saved Objects API response. If type isn't
+   * registered, response is returned as is.
+   * @param response Raw response returned by the underlying base client.
+   */
+  private stripEncryptedAttributesFromResponse<
+    T extends UpdateResponse | CreateResponse | GetResponse
+  >(response: T): T {
+    if (this.options.service.isRegistered(response.type)) {
+      response.attributes = this.options.service.stripEncryptedAttributes(
```

Comment:
__note:__ since it's response returned to the consumer I feel like it's okay to mutate existing object object here and in method below to make this operation a bit more efficient. But we never mutate `attributes` or `options` that are provided by the consumer since we can't be sure that these aren't used somewhere else.